README.md
## Assignment
Create a Content Management System as a Sinatra application.

## Running The Code
To run the program:
  1. download the program
  * run `$ bundle install` to install all dependencies
  * run `$ ruby twenty_one.rb`
  *  navigate to `localhost:4567` using a browser

To run rubocop:
  1. use version 0.46.0 (I was able to satisfy all but three small Rubocop complaints)

## Cross-browser Compatibility
The application works as expected on Chrome and FireFox.

## Baseline Features
1. User can view text documents (.txt, .md)
* User can rename or delete existing documents
* Styled Success and Error messages are displayed to User, list of files is formatted with a specific font and padding
* User can edit content of existing text documents
* There is a sign-in option and certain functionality (rename, delete) requires a user to be signed in
* User passwords are stored in a hashed format for security
* Tests exist for all functionality, including values stored in the session

## Additional Features
1. Only Markdown, MS Word Doc, and Text (.txt) files can be created natively within the application
* User can duplicate existing documents
* User can create a new account. New account creation signs the user in automatically, immediately allowing them to manipulate files
* User can upload images (.jpg/.jpeg, .png, .gif files) or PDF files and they will display properly


## Design Decisions
1. The `edit` link is now a button, and that functionality is only available for text documents.
* The `edit`, `duplicate`, `rename`, and `delete` buttons only display when a user is signed in.
* Only .doc, .md, and .txt files can be created natively. PDFs, images (.jpg, .jpeg, .gif, .png), or .md or .txt files can be uploaded. I was unable to render uploaded .doc files properly so those are not allowed.
* Renaming only allows the user to access the name of the file. This is in order to prevent possible corruption from inadvertent changes to the extension type.
* Uploaded files must be 1.5 MB or smaller.
* The user is asked to confirm their action when deleting a file.
* File names are transformed to CamelCase on save (creation, rename, or upload) and files are displayed alphabetically.
* Passwords are hidden from display.
* New user passwords are specified in the type of characters they must contain (start with an uppercase letter, contain at least one lowercase, one number and one special character: !\*-?)
* Every page other than the index features a `cancel` button to return the user to the index.
* Buttons are styled with hover effects to make the page slightly more interactive for the user. Success and Error messages are styled with different background colors for ease of disambiguation (though I worry that the red and green I chose are not ideal for those with red-green color blindness).


data/changes.txt
new content
testing

test/cms_test.rb
ENV['RACK_ENV'] = 'test'

require 'fileutils'
require 'minitest/autorun'
require 'minitest/pride'
require 'rack/test'

require_relative '../cms'

class CmsTest < Minitest::Test
  include Rack::Test::Methods

  def app
    Sinatra::Application
  end

  def setup
    FileUtils.mkdir_p(data_path)
  end

  def teardown
    FileUtils.rm_rf(data_path)
  end

  def session
    last_request.env['rack.session']
  end

  def create_document(name, content = '')
    File.open(File.join(data_path, name), 'w') do |file|
      file.write(content)
    end
  end

  def admin_session
    { 'rack.session' => { username: 'admin', signed_in: true } }
  end

  def random_username
    username = ''
    letters = ('a'..'z').to_a
    8.times do
      username += letters.sample
    end
    username
  end

  def test_index
    create_document('about.md')
    create_document('changes.txt')

    get '/'
    assert_equal(200, last_response.status)
    assert_equal('text/html;charset=utf-8', last_response['Content-Type'])
    assert_includes(last_response.body, 'About.md')
    assert_includes(last_response.body, 'Changes.txt')
  end

  def test_document
    create_document('history.txt', "1993 - Yukihiro Matsumoto dreams up Ruby.\n1995 - Ruby 0.95 released.")

    get '/history.txt'
    assert_equal(200, last_response.status)
    assert_equal 'text/plain', last_response['Content-Type']
    assert_includes(last_response.body, 'Yukihiro Matsumoto')
  end

  def test_document_not_found
    get '/bad_doc.erb'
    assert_equal(302, last_response.status)
    assert_equal('bad_doc.erb does not exist.', session[:error])

    get last_response['Location']
    assert_equal('text/html;charset=utf-8', last_response['Content-Type'])

    get '/'
    refute_equal('bad_doc.erb does not exist.', session[:error])
  end

  def test_markdown
    create_document('about.md', "* natural to read\n* easy to write")

    get '/about.md'
    assert_equal(200, last_response.status)
    assert_equal 'text/html', last_response['Content-Type']
    assert_includes(last_response.body, "<li>natural to read</li>\n<li>easy to write</li>")
  end

  def test_view_edit
    create_document('about.md')

    get '/about.md/edit', {}, admin_session
    assert_equal(200, last_response.status)
    assert_equal 'text/html;charset=utf-8', last_response['Content-Type']
    assert_includes(last_response.body, 'textarea')
    assert_includes(last_response.body, 'form')
    assert_includes(last_response.body, 'button')
  end

  def test_view_edit_not_signed_in
    get '/about.md/edit'
    assert_equal(302, last_response.status)
    assert_equal('You must be signed in to do that.', session[:error])
  end

  def test_post_edit
    create_document('changes.txt', 'old content')

    post '/changes.txt', {content: 'new content'}, admin_session
    assert_equal(302, last_response.status)
    assert_equal('changes.txt has been updated.', session[:success])

    get last_response['Location']
    assert_equal('text/html;charset=utf-8', last_response['Content-Type'])

    get '/'
    assert_equal(200, last_response.status)
    refute_equal('changes.txt has been updated.', session[:success])

    get '/changes.txt'
    assert_equal(200, last_response.status)
    assert_includes(last_response.body, 'new content')
  end

  def test_post_edit_not_signed_in
    create_document('changes.txt', 'old content')
    post '/changes.txt'
    assert_equal(302, last_response.status)
    assert_equal('You must be signed in to do that.', session[:error])
  end

  def test_new_doc_form
    get '/new', {}, admin_session
    assert_equal(200, last_response.status)
    assert_equal('text/html;charset=utf-8', last_response['Content-Type'])
    assert_includes(last_response.body, "<form action='/create' method='post'>")
    assert_includes(last_response.body, "<input type='text' name='file_name'")
    assert_includes(last_response.body, "<button class='new' type='submit'>")
  end

  def test_new_doc_form_not_signed_in
    get '/new'
    assert_equal(302, last_response.status)
    assert_equal('You must be signed in to do that.', session[:error])
  end

  def test_post_new_doc
    post '/create', {file_name: 'test.md'}, admin_session
    assert_equal(302, last_response.status)
    assert_equal('test.md was created.', session[:success])

    get '/'
    assert_equal(200, last_response.status)
    refute_equal('test.md was created.', session[:success] )
  end

  def test_post_new_doc_not_signed_in
    post '/create'
    assert_equal(302, last_response.status)
    assert_equal('You must be signed in to do that.', session[:error])
  end

  def test_post_invalid_doc_name
    post '/create', {file_name: 'test'}, admin_session
    assert_equal(422, last_response.status)
    assert_includes(last_response.body, 'Please include a valid extension')

    post '/create', {file_name: '   '}, admin_session
    assert_equal(422, last_response.status)
    assert_includes(last_response.body, 'A name is required.')
  end

  def test_post_new_doc_already_exists
    create_document('copy.txt')
    post '/create', {file_name: 'copy.txt'}, admin_session
    assert_equal(422, last_response.status)
    assert_includes(last_response.body, 'That file already exists.')
  end

  def test_post_rename
    create_document('Copy.txt')
    post '/Copy.txt/rename', {rename: 'Copy1.txt'}, admin_session
    assert_equal('Copy.txt was renamed to Copy1.txt.', session[:success])
    assert_equal(302, last_response.status)

    get last_response['Location']
    assert_equal(200, last_response.status)
    assert_includes(last_response.body, 'Copy1.txt</a>')
    refute_includes(last_response.body, 'Copy.txt</a>')
  end

  def test_post_rename_invalid_doc_name
    create_document('copy.txt')

    post '/copy.txt/rename', {rename: '   '}, admin_session
    assert_equal(422, last_response.status)
    assert_includes(last_response.body, 'A name is required.')
  end

  def test_post_rename_already_exists
    create_document('copy.txt')
    post '/copy.txt/rename', {rename: 'copy'}, admin_session
    assert_equal(422, last_response.status)
    assert_includes(last_response.body, 'That file already exists.')
  end

  def test_post_rename_not_signed_in
    post '/:file_name/rename'
    assert_equal(302, last_response.status)
    assert_equal('You must be signed in to do that.', session[:error])
  end

  def test_duplicate
    create_document('to_duplicate.md')
    post '/to_duplicate.md/duplicate', {}, admin_session
    assert_includes(session[:success], 'Duplication successful')
    assert_equal(302, last_response.status)

    get last_response['Location']
    assert_equal(200, last_response.status)
    assert_includes(last_response.body, 'to_duplicate_copy.md')
  end

  def test_duplicate_not_signed_in
    create_document('to_duplicate.md')
    post '/to_duplicate.md/duplicate'
    assert_equal(302, last_response.status)
    assert_equal('You must be signed in to do that.', session[:error])
  end

  def test_delete
    create_document('to_delete.md', '')
    post '/to_delete.md/delete', {file_name: 'to_delete.md'}, admin_session
    assert_equal(302, last_response.status)
    assert_equal('to_delete.md has been deleted.', session[:success])

    get last_response['Location']
    assert_equal('text/html;charset=utf-8', last_response['Content-Type'])

    get '/'
    assert_equal(200, last_response.status)
    refute_equal('to_delete.md has been deleted.', session[:success])
    refute_includes(last_response.body, 'to_delete.md')
  end

  def test_delete_not_signed_in
    create_document('to_delete.md', '')
    post '/to_delete.md/delete'
    assert_equal(302, last_response.status)
    assert_equal('You must be signed in to do that.', session[:error])
  end

  def test_display_registration_form
    get '/users/register'
    assert_equal(200, last_response.status)
    assert_equal('text/html;charset=utf-8', last_response['Content-Type'])
    assert_includes(last_response.body, "name='register'")
  end

  def test_registration
    username = random_username
    post '/users/register', new_username: username, new_password: 'A1!bcdef'
    assert_equal(302, last_response.status)
    assert_equal(true, session[:signed_in])
    assert_equal(username, session[:username])

    get last_response['Location']
    assert_includes(last_response.body, "Welcome, #{username}!")
  end

  def test_failed_registration
    post '/users/register', new_username: 'administrator', new_password: 'A1!bcdef'
    assert_equal(409, last_response.status)
    assert_includes(last_response.body, 'username already exists')

    post '/users/register', new_username: '', new_password: 'A1!bcdef'
    assert_equal(422, last_response.status)
    assert_includes(last_response.body, 'Please enter a valid username and password.')

    post '/users/register', new_username: 'testingnames', new_password: ''
    assert_equal(422, last_response.status)
    assert_includes(last_response.body, 'Please enter a valid username and password.')
  end

  def test_sign_in_form
    get '/users/signin'
    assert_equal(200, last_response.status)
    assert_equal('text/html;charset=utf-8', last_response['Content-Type'])
    assert_includes(last_response.body, "id='username'")
    assert_includes(last_response.body, "id='password'")
    assert_includes(last_response.body, "name='signin'>Sign In</button>")
  end

  def test_sign_in
    post '/users/signin', username: 'administrator', password: 'A1!bcdef'
    assert_equal(302, last_response.status)
    assert_equal('Welcome!', session[:success])

    get last_response['Location']
    assert_equal('text/html;charset=utf-8', last_response['Content-Type'])

    assert_includes(last_response.body, 'Signed in as')
    assert_includes(last_response.body, 'Sign Out')

    get '/'
    assert_equal(200, last_response.status)
    assert_includes(last_response.body, 'Signed in as')
    refute_equal('Welcome!', session[:success])
    assert_includes(last_response.body, 'Sign Out')
  end

  def test_failed_sign_in
    post '/users/signin', username: '', password: ''
    assert_equal(409, last_response.status)
    assert_includes(last_response.body, 'Invalid Credentials')

    post '/users/signin', username: 'xxx', password: 'xxx'
    assert_equal(409, last_response.status)
    assert_includes(last_response.body, 'Invalid Credentials')
  end

  def test_sign_out
    get '/', {}, admin_session
    assert_includes(last_response.body, 'Signed in as admin')

    post '/users/signout'
    assert_equal(302, last_response.status)
    assert_equal('You have been signed out.', session[:success])

    get last_response['Location']
    assert_nil(session[:username])
    assert_includes(last_response.body, 'Sign In')

    get '/'
    assert_equal(200, last_response.status)
    refute_equal('You have been signed out.', session[:success])
    refute_includes(last_response.body, 'Signed in as')
  end

  def test_display_upload_form
    get '/upload', {}, admin_session
    assert_equal(200, last_response.status)
    assert_includes(last_response.body, 'enctype="multipart/form-data"')
  end

  def test_display_upload_form_not_signed_in
    get '/upload'
    assert_equal('You must be signed in to do that.', session[:error])
    assert_equal(302, last_response.status)
  end

  def test_upload
    file_path = File.join(upload_path, 'panda.gif')
    file = Rack::Test::UploadedFile.new(file_path, 'image/jpeg')

    post '/upload', {fileupload: file}, admin_session
    assert_equal('Panda.gif was uploaded.', session[:success])

    get last_response['Location']
    assert_includes(last_response.body, 'Panda.gif')
  end

  def test_upload_duplicate_file
    file_path = File.join(upload_path, 'panda.gif')
    file = Rack::Test::UploadedFile.new(file_path, 'image/jpeg')

    post '/upload', {fileupload: file}, admin_session
    post '/upload', {fileupload: file}, admin_session
    assert_equal(422, last_response.status)
    assert_includes(last_response.body, 'That file already exists.')
  end

  def test_upload_no_file
    post '/upload', {}, admin_session
    assert_equal(422, last_response.status)
    assert_includes(last_response.body, 'Please select a file to upload.')
  end

  def test_upload_unsupported_file
    file_path = File.join(upload_path, 'test.doc')
    file = Rack::Test::UploadedFile.new(file_path, 'text/plain')

    post '/upload', {fileupload: file}, admin_session
    assert_equal(422, last_response.status)
    assert_includes(last_response.body, 'That file type is unsupported.')
  end

  def test_upload_file_too_large
    file_path = File.join(upload_path, 'TangerineGimlet.JPG')
    file = Rack::Test::UploadedFile.new(file_path, 'image/png')

    post '/upload', {fileupload: file}, admin_session
    assert_equal(422, last_response.status)
    assert_includes(last_response.body, 'The file is too big. Please resize')
  end

  def test_upload_not_signed_in
    post '/upload'
    assert_equal('You must be signed in to do that.', session[:error])
    assert_equal(302, last_response.status)
  end
end


cms.rb
require 'bcrypt'
require 'redcarpet'
require 'sinatra'
require 'sinatra/reloader'
require 'sinatra/content_for'
require 'tilt/erubis'
require 'yaml'

TEXT_EXTENSIONS = %w(.md .txt .doc).freeze
IMAGE_EXTENSIONS = %w(.jpg .jpeg .gif .png).freeze
UPLOAD_EXTENSIONS = %w(.md .txt .pdf .jpg .jpeg .gif .png).freeze

configure do
  enable :sessions
  set :session_secret, 'secret'
  set :erb, escape_html: true
end

before do
  pattern = File.join(data_path, '*')
  @files = Dir.glob(pattern).map do |path|
    name = File.basename(path)
    name.start_with?(/[A-Z]/) ? name : name.capitalize
  end.sort
end

helpers do
  def read_message
    if session[:error]
      :error
    elsif session[:success]
      :success
    end
  end

  def split_name(file_name)
    file_name.split('.')
  end

  def text_extension?(file_name)
    TEXT_EXTENSIONS.include?(File.extname(file_name))
  end
end

def data_path
  if ENV['RACK_ENV'] == 'test'
    File.expand_path('../test/data', __FILE__)
  else
    File.expand_path('../data', __FILE__)
  end
end

def upload_path
  File.expand_path('../test', __FILE__) if ENV['RACK_ENV'] == 'test'
end

def credentials_path
  if ENV['RACK_ENV'] == 'test'
    File.expand_path('../test/users.yml', __FILE__)
  else
    File.expand_path('../users.yml', __FILE__)
  end
end

def load_user_credentials
  YAML.load_file(credentials_path)
end

def encrypt_password(password)
  BCrypt::Password.create(password)
end

def clean_yaml
  content = ''
  File.open(credentials_path) do |users|
    content = users.read.gsub(/\n-{3}/, '')
  end
  File.open(credentials_path, 'w') do |users|
    users.write(content)
  end
end

# append new user to credentials list
def write_user_credentials(username, password)
  File.open(credentials_path, 'a') do |users|
    users.write(Psych.dump("#{username}": encrypt_password(password)))
  end
  clean_yaml
end

def invalid_file_type?(possible_extensions, file_name)
  !possible_extensions.include?(File.extname(file_name).downcase)
end

def simplify_file_name!(name)
  file, extension = split_name(name)
  extension = extension.gsub(/[A-Z]/, 'a-z')
  file = file.split.map(&:capitalize).join.gsub(/['"]/, '').strip
  "#{file}.#{extension}"
end

def render_markdown(text)
  markdown = Redcarpet::Markdown.new(Redcarpet::Render::HTML)
  markdown.render(text)
end

def verify_signed_in
  return if session[:signed_in]
  session[:error] = 'You must be signed in to do that.'
  redirect '/'
end

def valid_credentials?(username, password)
  credentials = load_user_credentials
  if credentials.key?(username.to_sym)
    bcrypt_password = BCrypt::Password.new(credentials[username])
    bcrypt_password == password
  else
    false
  end
end

def determine_text_display(extension, content)
  case extension
  when '.txt', '.doc'
    headers['Content-Type'] = 'text/plain'
    content
  when '.md'
    headers['Content-Type'] = 'text/html'
    erb render_markdown(content)
  end
end

def load_file_content(path)
  content = File.read(path)
  extension = File.extname(path)
  if TEXT_EXTENSIONS.include?(extension)
    determine_text_display(extension, content)
  elsif IMAGE_EXTENSIONS.include?(extension)
    headers['Content-Type'] = 'image/jpg'
    content
  elsif extension == '.pdf'
    headers['Content-Type'] = 'application/pdf'
    content
  end
end

def create_document(name, content = '')
  File.open(File.join(data_path, name), 'w') do |file|
    file.write(content)
  end
end

def file_exists?(file_name)
  File.file?(File.join(data_path, file_name))
end

def file_too_large?(file_name)
  File.size(file_name) >= 1_500_000
end

def invalid_new_document?(doc_name)
  if doc_name.empty?
    'A name is required.'
  elsif invalid_file_type?(TEXT_EXTENSIONS, doc_name)
    'Please include a valid extension for your file (use ' \
    "#{TEXT_EXTENSIONS.join(', ')})."
  elsif file_exists?(simplify_file_name!(doc_name))
    'That file already exists. Please choose another name.'
  end
end

def invalid_rename?(new_name, extension)
  if new_name.empty?
    'A name is required.'
  elsif file_exists?("#{new_name}.#{extension}")
    'That file already exists. Please choose another name.'
  end
end

def invalid_upload?(file_details)
  if file_details.nil?
    'Please select a file to upload.'
  elsif invalid_file_type?(UPLOAD_EXTENSIONS, file_details[:filename])
    "Unsupported file type. Please only use #{UPLOAD_EXTENSIONS.join(', ')}."
  elsif file_exists?(file_details[:filename])
    'That file already exists.'
  elsif file_too_large?(file_details[:tempfile])
    'The file is too big. Please resize or try another file.'
  end
end

# load index
get '/' do
  erb :index
end

# display registration form
get '/users/register' do
  erb :register
end

# register new user
post '/users/register' do
  username = params[:new_username]
  password = params[:new_password]
  if username.empty? || password.empty?
    status 422
    session[:error] = 'Please enter a valid username and password.'
    erb :register
  elsif load_user_credentials.key?(username.to_sym)
    status 409
    session[:error] = 'That username already exists. Please choose another.'
    erb :register
  else
    write_user_credentials(username, password)
    session[:success] = "Account successfully registered. Welcome, " \
                        "#{username}! Please save your password for future " \
                        "refrerence: #{password}"
    session[:signed_in] = true
    session[:username] = username
    redirect '/'
  end
end

# display form to create new document
get '/new' do
  verify_signed_in
  erb :new_file, layout: :layout
end

# display sign in form
get '/users/signin' do
  erb :sign_in
end

# sign in
post '/users/signin' do
  username = params[:username].strip
  password = params[:password].strip
  if valid_credentials?(username.to_sym, password)
    session[:success] = 'Welcome!'
    session[:signed_in] = true
    session[:username] = username
    redirect '/'
  else
    session[:error] = 'Invalid Credentials'
    status 409
    erb :sign_in
  end
end

# sign out
post '/users/signout' do
  session[:signed_in] = false
  session.delete(:username)
  session[:success] = 'You have been signed out.'
  redirect '/'
end

# display upload form
get '/upload' do
  verify_signed_in
  erb :upload
end

# upload file
post '/upload' do
  verify_signed_in
  file_details = params[:fileupload]
  session[:error] = invalid_upload?(file_details)
  if session[:error]
    status 422
    erb :upload
  else
    file_name = simplify_file_name!(file_details[:filename])
    file = file_details[:tempfile]
    FileUtils.mv(file, File.join(data_path, file_name))
    session[:success] = "#{file_name} was uploaded."
    redirect '/'
  end
end

# create new document
post '/create' do
  verify_signed_in
  doc_name = params[:file_name].strip
  session[:error] = invalid_new_document?(doc_name)
  if session[:error]
    status 422
    erb :new_file
  else
    create_document(doc_name)
    session[:success] = "#{doc_name} was created."
    redirect '/'
  end
end

# display rename form
get '/:file_name/rename' do
  verify_signed_in
  erb :rename
end

# rename document
post '/:file_name/rename' do
  verify_signed_in
  old_name = params[:file_name]
  extension = split_name(params[:file_name]).last
  new_name = params[:rename].strip
  session[:error] = invalid_rename?(new_name, extension)
  if session[:error]
    status 422
    erb :rename
  else
    new_name = simplify_file_name!("#{new_name}.#{extension}")
    File.rename(File.join(data_path, old_name), File.join(data_path, new_name))
    session[:success] = "#{old_name} was renamed to #{new_name}."
    redirect '/'
  end
end

# duplicate a document
post '/:file_name/duplicate' do
  verify_signed_in
  doc_name = params[:file_name]
  name, extension = split_name(doc_name)
  content = File.read(File.join(data_path, doc_name))
  duplicate_name = name + '_copy.' + extension
  create_document(duplicate_name, content)
  session[:success] = "Duplication successful: #{duplicate_name} created."
  redirect '/'
end

# delete a document
post '/:file_name/delete' do
  verify_signed_in
  doc = params[:file_name]
  file_path = File.join(data_path, doc)
  if env['HTTP_X_REQUESTED_WITH'] == 'XMLHttpRequest'
    File.delete(file_path)
    status 204
  else
    File.delete(file_path)
    session[:success] = "#{doc} has been deleted."
    redirect '/'
  end
end

# display document
get '/:file_name' do
  doc = params[:file_name]
  file_path = File.join(data_path, doc)
  if File.file?(file_path)
    @document = load_file_content(file_path)
  else
    session[:error] = "#{doc} does not exist."
    redirect '/'
  end
end

# display edit form
get '/:file_name/edit' do
  verify_signed_in
  @doc = params[:file_name]
  file_path = File.join(data_path, @doc)
  @document = File.read(file_path)
  erb :edit
end

# submit edits to file
post '/:file_name' do
  verify_signed_in
  doc = params[:file_name]
  file_path = File.join(data_path, doc)
  File.write(file_path, params[:content])
  session[:success] = "#{doc} has been updated."
  redirect '/'
end
